recipe/2024-02-terraform-providers.md

```diff
@@ -0,0 +1,690 @@
+
+### User scenarios (optional)
+
+#### User story 1
+As an operator, I maintain a set of Terraform recipes for use with Radius. I have a set of provider configurations which are applicable across multiple recipes and I would like to configure them in a single centralized location for ease of maintenance.
+
+
+#### User story 2
+As an operator, I would like to manage cloud resources in different geographical regions. To enable this, I need to set up multiple configurations for the same provider for different regions and use an alias to refer to individual provider configurations. 
+
+## Design
+### Design details
+
+Reviewing popular provider configurations (GCP, Oracle, Heroku, DigitalOcean, Docker etc.) a large number of provider configurations can be setup by handling a combination of key value pairs, key value pairs in nested objects, secrets and environment variables.
+
+### Key-value pairs
+Attributes or settings, represented as key-value pairs are often used to define configuration parameters in provider configurations.
+We will parse and save these values in the Terraform configuration file, main.tf in a working directory, which the system creates today. 
+
+### Alias
+
+Aliases in Terraform allow users to manage multiple instances of the same provider with different configurations for a single module. 
+Ref links: 
+- [Alias Multiple Provider Configurations](https://developer.hashicorp.com/terraform/language/providers/configuration#alias-multiple-provider-configurations)
+- [Passing Providers Explicitly](https://developer.hashicorp.com/terraform/language/modules/develop/providers#passing-providers-explicitly)
+
+For a Terraform recipe, aliases needed by the module are declared in the required_providers block, where each alias is a unique provider configuration.
+eg.
+
+```
+terraform {
+  required_providers {
+    mycloud = {
+      source  = "mycorp/mycloud"
+      version = "~> 1.0"
+      configuration_aliases = [ mycloud.west_region, mycloud.east_region ]
+    }
+  }
+}
+// The `required_providers` block for this module specifies that it will need provider configuration for aliases west_region and east_region to deploy successfully.
+```
+
+The bicep configuration for Terraform providers described in the environment resource is typically written to `provider` block in the main.tf.json file. These provider configurations are then passed to the recipe module. For provider configurations without aliases, they are be passed to the module automatically by Terraform.
+
+However, when the recipe module specifies one or more aliases in the `required_providers` block (and the module does not contain the provider configuration details), the aliased provider configuration from the `provider` block needs to be passed explicitly to the module.
+
+This is done by updating the module block in the main.tf.json file with provider aliases.
+
```

Comment:
It would be helpful to me to understand what need to *change* to support aliases.

- I think we generate `required_providers`. It sounds like that changes.
- We generate a `provider` block for each provider/alias. I think it's new that we're aware of aliases in this code.
- We generate a call to the module, it sounds like this needs to change too.

Reply:
I think going over the example from below in more detail (what's new) will answer my question.